### PR TITLE
JAMES-2228 Total & unread messages numbers should never be negative

### PR DESCRIPTION
**Pull request by aduprat**

### CODE REVIEW

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/mailbox/Mailbox.java [hunk 1]
```diff
@@ -166,7 +168,12 @@ public Mailbox build() {
             Preconditions.checkState(id != null, "'id' is mandatory");
 
             return new Mailbox(id, name, parentId, role, sortOrder, mustBeOnlyMailbox, mayReadItems, mayAddItems, mayRemoveItems, mayCreateChild, mayRename, mayDelete,
-                    totalMessages, unreadMessages, totalThreads, unreadThreads, sharedWith.orElse(Rights.EMPTY), namespace.orElse(MailboxNamespace.personal()));
+                    negativeToZero(totalMessages), negativeToZero(unreadMessages), totalThreads, unreadThreads, sharedWith.orElse(Rights.EMPTY), namespace.orElse(MailboxNamespace.personal()));
+        }
+
+        private long negativeToZero(Optional<Long> number) {
+            return number.filter(value -> value >= 0)
+                .orElse(0l);
```
Comment (chibenwa) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/model/mailbox/Mailbox.java]: 0L overwize I have a warning in intelliJ

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/model/mailbox/MailboxTest.java [hunk 2]
```diff
@@ -96,4 +96,26 @@ public void parentIdDefaultValueIsEmpty() {
 
         assertThat(mailbox.getParentId()).isEmpty();
     }
+
+    @Test
```
Comment (chibenwa) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/model/mailbox/MailboxTest.java]: You miss tests with value 0 ?

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/model/mailbox/MailboxTest.java [hunk 3]
```diff
@@ -118,4 +118,48 @@ public void unreadMessagesShouldNeverBeNegative() {
 
         assertThat(mailbox.getUnreadMessages()).isEqualTo(0);
     }
+
+    @Test
+    public void totalMessagesShouldZeroWhenZero() {
```
Comment (rouazana) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/model/mailbox/MailboxTest.java]: s/Should/ShouldReturn/

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/model/mailbox/MailboxTest.java [hunk 4]
```diff
@@ -118,4 +118,48 @@ public void unreadMessagesShouldNeverBeNegative() {
...
+                .id(InMemoryId.of(1))
+                .name("name")
+                .totalMessages(0)
+                .build();
+
+        assertThat(mailbox.getTotalMessages()).isEqualTo(0);
+    }
+
+    @Test
+    public void unreadMessagesShouldZeroWhenZero() {
```
Comment (rouazana) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/model/mailbox/MailboxTest.java]: s/Should/ShouldReturn/

**Comment by chibenwa** [comment]
```
[9086baef33764a005ce28cc7f77bf9bec7a5657b] [INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 21.899 s - in org.apache.james.jmap.cassandra.CassandraUserProvisionningConcurrencyTest
```

test this please

**Comment by chibenwa** [comment]
```
[9086baef33764a005ce28cc7f77bf9bec7a5657b] [ERROR] Errors: 
[9086baef33764a005ce28cc7f77bf9bec7a5657b] [ERROR]   CassandraSetMessagesMethodTest>SetMessagesMethodTest.attachmentsAndBodyShouldBeRetrievedWhenChainingSetMessagesAndGetMessagesWithTextBodyAndHtmlAttachment:3900 » ConditionTimeout
```

**Comment by aduprat** [comment]
test this please

**Comment by aduprat** [comment]
```
[ERROR] org.apache.maven.surefire.booter.SurefireBooterForkException: ExecutionException The forked VM terminated without properly saying goodbye. VM crash or System.exit called?
[ERROR] Command was /bin/sh -c cd /james-project/server/protocols/jmap-integration-testing/cassandra-jmap-integration-testing && /usr/lib/jvm/java-8-openjdk-amd64/jre/bin/java -Xms256m -Xmx512m -jar /james-project/server/protocols/jmap-integration-testing/cassandra-jmap-integration-testing/target/surefire/surefirebooter913889045241107535.jar /james-project/server/protocols/jmap-integration-testing/cassandra-jmap-integration-testing/target/surefire 2017-11-23T08-29-44_164-jvmRun1 surefire7849714872979191256tmp 
```
test this please

**Comment by chibenwa** [comment]
test this please

**Comment by aduprat** [comment]
merged

**Pull request closed without merging**